tests/ProtoTests/ProtoTest.php

```diff
@@ -27,181 +27,27 @@ final class ProtoTest extends TestCase
-
-        // Check generator code is as expected.
-        $all = Vector::new($codeIterator);
-
-        // TODO: Handle multiple files in a general way.
-        // TODO: Move expected text to files, not inline in the test code.
-
-        // Check GAPIC client generation
-        $expectedFilename = 'Gapic/BasicGapicClient.php';
-        [$filename, $code] = $all->filter(fn($x) => $x[0] === $expectedFilename)[0];
-        $expectedCode = <<<'EOF'
-<?php
-
-declare(strict_types=1);
-
-namespace testing\basic\Gapic;
-
-use Google\ApiCore\CredentialsWrapper;
-use Google\ApiCore\GapicClientTrait;
-use Google\ApiCore\Transport\GrpcTransport;
-use Google\ApiCore\Transport\RestTransport;
-use Google\ApiCore\Transport\TransportInterface;
-use Google\ApiCore\ValidationException;
-use Google\Auth\FetchAuthTokenInterface;
-
-/**
- * Service Description: This is a basic service.
- *
- * This class provides the ability to make remote calls to the backing service through method
- * calls that map to API methods. Sample code to get started:
- *
- * @experimental
- */
-class BasicGapicClient
-{
-    use GapicClientTrait;
-
-    /** The name of the service. */
-    const SERVICE_NAME = 'testing.basic.Basic';
-
-    /** The default address of the service. */
-    const SERVICE_ADDRESS = 'basic.example.com';
-
-    /** The default port of the service. */
-    const DEFAULT_SERVICE_PORT = 443;
-
-    /** The name of the code generator, to be included in the agent header. */
-    const CODEGEN_NAME = 'gapic';
-
-    /** The default scopes required by the service. */
-    public static $serviceScopes = [
-        'scope1',
-        'scope2',
-    ];
-
-    private static function getClientDefaults()
-    {
-        return [
-            'serviceName' => self::SERVICE_NAME,
-            'apiEndpoint' => self::SERVICE_ADDRESS . ':' . self::DEFAULT_SERVICE_PORT,
-            'clientConfig' => __DIR__ . '/../resources/basic_client_config.json',
-            'descriptorsConfigPath' => __DIR__ . '/../resources/basic_descriptor_config.php',
-            'gcpApiConfigPath' => __DIR__ . '/../resources/basic_grpc_config.json',
-            'credentialsConfig' => [
-                'scopes' => self::serviceScopes,
-            ],
-            'transportConfig' => [
-                'rest' => [
-                    'restClientConfigPath' => __DIR__ . '/../resources/basic_rest_client_config.php',
-                ],
-            ],
-        ];
+        // Conventions:
+        // * The proto package is 'testing.<proto-name>'.
+        // * The expected file contents are based in the same directory as the proto file.
+        $descBytes = $this->loadDescriptorBytes("ProtoTests/{$protoPath}");
+        $package = 'testing.' . basename($protoPath, '.proto');
+        $codeIterator = CodeGenerator::GenerateFromDescriptor($descBytes, $package);
+
+        foreach ($codeIterator as [$relativeFilename, $code]) {
```

Comment:
One nit: an empty code generator would still pass this test, because we don't check that all the expected files are actually generated.

Reply:
Good point, thanks. TODO added.